lib/clientv2/node.go

```
@@ -59,7 +63,103 @@ func (r nodes) Update(ctx context.Context, res *apiv2.Node, opts options.SetOpti
+	switch err.(type) {
+	case nil, errors.ErrorResourceDoesNotExist, errors.ErrorOperationNotSupported:
+	default:
+		return nil, err
+	}
+
+	// Delete the weps.
+	for _, wep := range weps.Items {
+		if wep.Spec.Node != name {
+			continue
+		}
+
+		_, err = r.client.WorkloadEndpoints().Delete(ctx, wep.Namespace, wep.Name, options.DeleteOptions{})
+		switch err.(type) {
+		case nil, errors.ErrorResourceDoesNotExist, errors.ErrorOperationNotSupported:
+		default:
+			return nil, err
+		}
+	}
+
+	// Remove the node from the IPAM data if it exists.
+	err = r.client.IPAM().RemoveIPAMHost(ctx, name)
+	switch err.(type) {
+	case nil, errors.ErrorResourceDoesNotExist, errors.ErrorOperationNotSupported:
+	default:
+		return nil, err
+	}
+
+	// Remove BGPPeers.
+	bgpPeers, err := r.client.BGPPeers().List(ctx, options.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	for _, peer := range bgpPeers.Items {
+		if peer.Spec.Node != name {
+			continue
+		}
+		_, err = r.client.BGPPeers().Delete(ctx, peer.Name, options.DeleteOptions{})
+		switch err.(type) {
+		case nil, errors.ErrorResourceDoesNotExist, errors.ErrorOperationNotSupported:
+		default:
+			return nil, err
+		}
+	}
+
+	// Delete felix configuration
+	nodeConfName := fmt.Sprintf("node.%s", name)
+	_, err = r.client.FelixConfigurations().Delete(ctx, nodeConfName, options.DeleteOptions{})
+	switch err.(type) {
+	case nil, errors.ErrorResourceDoesNotExist, errors.ErrorOperationNotSupported:
+	default:
+		return nil, err
+	}
+
+	// Delete bgp configuration
+	_, err = r.client.BGPConfigurations().Delete(ctx, nodeConfName, options.DeleteOptions{})
+	switch err.(type) {
+	case nil, errors.ErrorResourceDoesNotExist, errors.ErrorOperationNotSupported:
+	default:
+		return nil, err
+	}
+
+	// Delete the node.
 	out, err := r.client.resources.Delete(ctx, opts, apiv2.KindNode, noNamespace, name)
+	switch err.(type) {
+	case nil, errors.ErrorResourceDoesNotExist, errors.ErrorOperationNotSupported:
```

When deleting the actual node we should return the ResourceDoesNotExist error if there was no per-node configuration to delete at all.

The OperationNotSupported is more interesting as we may have deleted some other config.  @caseydavenport - what do you reckon for this error - perhaps return a not supported if no other config was deleted?

---

Looks like we didn't swallow any errors on the v1 api: https://github.com/projectcalico/libcalico-go/blob/master/lib/client/node.go#L80-L132